lib/assets/javascripts/turbograft/turbohead.coffee

```diff
@@ -76,37 +101,37 @@ noDatasetMatchesIn = (attribute, collection) ->
 updateLinkTags = (activeDocument, newLinks) ->
   # style tag load events don't work in all browsers
   # as such we just hope they load ¯\_(ツ)_/¯
-  newLinks.forEach((linkNode) -> insertLinkTask(activeDocument, linkNode)())
-
-updateScriptTags = (activeDocument, newScripts, callback) ->
-  asyncSeries(
-    newScripts.map((scriptNode) -> insertScriptTask(activeDocument, scriptNode)),
-    callback
+  newLinks.forEach((linkNode) ->
+    newNode = linkNode.cloneNode()
+    activeDocument.head.appendChild(newNode)
+    triggerEvent("page:after-link-inserted", newNode)
   )
 
-asyncSeries = (tasks, callback) ->
-  return callback() if tasks.length == 0
-  task = tasks.shift()
-  task(-> asyncSeries(tasks, callback))
+updateScriptTags = (activeDocument, newScripts) ->
+  promise = Promise.resolve()
+  newScripts.forEach (scriptNode) ->
+    promise = promise.then(-> insertScript(activeDocument, scriptNode))
+  promise
+
+insertScript = (activeDocument, scriptNode) ->
+  url = scriptNode.src
+  if scriptPromises[url]
+    return scriptPromises[url]
 
-insertScriptTask = (activeDocument, scriptNode) ->
-  # We need to clone script tags in order to ensure that the browser executes them.
+  # Clone script tags to guarantee browser execution.
   newNode = activeDocument.createElement('SCRIPT')
   newNode.setAttribute(attr.name, attr.value) for attr in scriptNode.attributes
   newNode.appendChild(activeDocument.createTextNode(scriptNode.innerHTML))
```

Comment:
I'm unclear how this part works. We clone the script tags here, but then also again in [turbolinks.coffee](https://github.com/Shopify/turbograft/blob/b639629df256fcd1045b6665af75ec4d5fffd9fa/lib/assets/javascripts/turbograft/turbolinks.coffee#L285-L292)? In this case the second clone only happens if `isCanceled` is true, but won't that mean they get executed once or twice, rather than never or once?